collections/src/main/scala/strawman/collection/mutable/ArrayBuffer.scala

```diff
@@ -40,48 +40,55 @@ class ArrayBuffer[A] private (initElems: Array[AnyRef], initLength: Int)
 
   def this() = this(new Array[AnyRef](16), 0)
 
-  def this(initLength: Int) = this(new Array[AnyRef](initLength), 0)
+  def this(initSize: Int) = this(new Array[AnyRef](initSize), 0)
 
   private var array: Array[AnyRef] = initElems
-  private var end = initLength
+  private var size0 = initSize
 
   /** Ensure that the internal array has at least `n` cells. */
   private def ensureSize(n: Int): Unit =
-    array = RefArrayUtils.ensureSize(array, end, n)
+    array = RefArrayUtils.ensureSize(array, size0, n)
 
   /** Reduce length to `n`, nulling out all dropped elements */
   private def reduceToSize(n: Int): Unit = {
-    RefArrayUtils.nullElems(array, n, end)
-    end = n
+    RefArrayUtils.nullElems(array, n, size0)
+    size0 = n
   }
 
-  private def checkWithinBounds(lo: Int, hi: Int) = {
+  @inline private def checkWithinBounds(lo: Int, hi: Int) = {
     if (lo < 0) throw new IndexOutOfBoundsException(lo.toString)
-    if (hi > end) throw new IndexOutOfBoundsException(hi.toString)
+    if (hi > size0) throw new IndexOutOfBoundsException(hi.toString)
   }
 
   @throws[ArrayIndexOutOfBoundsException]
-  def apply(n: Int) = array(n).asInstanceOf[A]
+  def apply(n: Int) = {
+    checkWithinBounds(n, n + 1)
+    array(n).asInstanceOf[A]
```

Comment:
Aren’t bounds already checked by the underlying array access?

Reply:
the underlying array can be larger than the `ArrayBuffer` in many situations:

```
import strawman.collection.mutable.ArrayBuffer
val a = ArrayBuffer(1)
a: strawman.collection.mutable.ArrayBuffer[Int] = ArrayBuffer(1)

scala> a.remove(0)
res2: Int = 1

scala> a(0)
res3: Int = 0

scala> val b = ArrayBuffer.empty[Int]
b: strawman.collection.mutable.ArrayBuffer[Int] = ArrayBuffer()

scala> b(0)
res4: Int = 0

scala> val c = ArrayBuffer(1, 2, 3)
c: strawman.collection.mutable.ArrayBuffer[Int] = ArrayBuffer(1, 2, 3)

scala> c.clear()

scala> c.size
res6: Int = 0

scala> c(0)
res7: Int = 1
```

Reply:
in summary, when an `ArrayBuffer` is created empty, its underlying array's size will be 16. When an `ArrayBuffer` is shrunken (remove/clear), the backed array remains with the same size